src/compiler/crystal/semantic/top_level_visitor.cr

```diff
@@ -70,14 +86,30 @@ class Crystal::TopLevelVisitor < Crystal::SemanticVisitor
       end
     else
       created_new_type = true
-      if type_vars = node.type_vars
-        type = GenericClassType.new @program, scope, name, nil, type_vars, false
-        type.splat_index = node.splat_index
-      else
-        type = NonGenericClassType.new @program, scope, name, nil, false
+      case special_type
+      in Nil
+        if type_vars = node.type_vars
+          type = GenericClassType.new @program, scope, name, nil, type_vars, false
+          type.splat_index = node.splat_index
+        else
+          type = NonGenericClassType.new @program, scope, name, nil, false
+        end
+        type.abstract = node.abstract?
+        type.struct = node.struct?
+      in .reference_storage_type?
+        unless type_vars = node.type_vars
+          node.raise "BUG: Expected reference_storage_type to be a generic struct"
+        end
+        unless type_vars.size == 1
+          node.raise "BUG: Expected reference_storage_type to have a single generic type parameter"
+        end
+        if node.splat_index
+          node.raise "BUG: Expected reference_storage_type to have no splat parameter"
+        end
+        type = GenericReferenceStorageType.new @program, scope, name, @program.value, type_vars
```

Comment:
**chore:** Can we also validate `node.struct? == true` and `node.abstract? == false`?
`abstract class ReferenceStorage(T)` should be invalid.